rclcpp/src/rclcpp/executors/executor_notify_waitable.cpp

```diff
@@ -99,22 +99,73 @@ ExecutorNotifyWaitable::take_data()
   return nullptr;
 }
 
+std::shared_ptr<void>
+ExecutorNotifyWaitable::take_data_by_entity_id(size_t id)
+{
+  (void) id;
+  return nullptr;
+}
+
+void
+ExecutorNotifyWaitable::set_on_ready_callback(std::function<void(size_t, int)> callback)
+{
+  // The second argument of the callback should identify which guard condition
+  // triggered the event.
+  // We could indicate which of the guard conditions was triggered, but the executor
+  // is already going to check that.
```

Comment:
Actually the executor doesn't check which guard condition was triggered, it just knows the waitable to whom it belongs.

If the executor knows which `gc` was triggered, it could more efficiently do work on the node or entity having this `gc`.
For example right know, a single `gc` triggered due addition of an entity to a node, will result on re-setting the callbacks to all nodes entities registered in the executor. We could use this info to get only the node having this `gc` and setting the callbacks to it. This is just a peformance improvement, but we already have the `notify_waitable_event_pushed_` which is mitigating this issue. 
Also if we manage to add apis to not notify the executor for every entity added, until manually done later, this will be less a problem

Reply:
Few comments here:
 1. Yes, the executor doesn't check which guard condition triggered, but we could easily do that (maybe I should change the `should` into `could` in the comment)
 2. This new implementation does not re-set the callbacks to all the nodes. It performs a loop on all the entities to understand which ones have been added/removed and then it does only the necessary work. By taking into account the possible improvement mentioned in (1) we could speed up the loop (i.e. loop only on one node, rather than on all nodes), but the set/unset would remain equivalent.

IMO, the fact that we don't reset everything, plus the presence of the `notify_waitable_event_pushed_` flag, should already improve the resource usage during startup.

Controlling the notification to the executor would still be important to reduce the amount of time we loop over the entities.

Reply:
> Also if we manage to add apis to not notify the executor for every entity added, until manually done later, this will be less a problem

While we can't get this for Iron, I would like to get it in sooner rather than later in rolling while we all still have context.